### PR TITLE
stty: Implement printing assignment of control chars

### PR DESCRIPTION
**Pull request by dezgeg**
Part of #3861

### CODE REVIEW

File: src/uu/stty/src/stty.rs [hunk 1]
```diff
@@ -245,6 +245,42 @@ fn print_terminal_size(termios: &Termios, opts: &Options) -> nix::Result<()> {
...
+        cc &= !0x80;
+        prefix = "M-";
+    }
+
+    match cc {
+        0..=0x1f => Ok(format!("{}^{}", prefix, (b'@' + cc) as char)),
+        0x20..=0x7e => Ok(format!("{}{}", prefix, cc as char)),
+        0x7f => Ok(format!("{}^?", prefix)),
+        _ => Err(nix::errno::Errno::ERANGE),
+    }
```
Comment (tertsdiepraam) [inline src/uu/stty/src/stty.rs:262]: I tried to clean this up a bit. What do you think of this?
```suggestion
    if cc == 0 {
        return Ok("<undef>".to_string());
    }

    let (prefix, code) = if cc >= 0x80 {
        ("M-", cc - 0x80)
    } else {
        ("", cc)
    };

    let (ctrl, character) = match code {
        0..=0x1f => Ok(("^", (b'@' + code) as char)),
        0x20..=0x7e => Ok(("", code as char)),
        0x7f => Ok(("^", '?')),
        _ => Err(nix::errno::Errno::ERANGE),
    }?;

    Ok(format!("{prefix}{ctrl}{character}"))
```

File: src/uu/stty/src/stty.rs [hunk 2]
```diff
@@ -245,6 +245,42 @@ fn print_terminal_size(termios: &Termios, opts: &Options) -> nix::Result<()> {
...
+}
+
+fn print_control_chars(termios: &Termios, opts: &Options) -> nix::Result<()> {
+    if !opts.all {
```
Comment (tertsdiepraam) [inline src/uu/stty/src/stty.rs:266]: Are control chars never printed? Not even if they're changed? It's fine if this is temporary, but then there should be a TODO in the code here probably.

**Comment by tertsdiepraam** [comment]
Very cool to see `stty` get so much attention :) Nice work!

File: src/uu/stty/src/stty.rs (hunk 2 above)
Comment (dezgeg) [reply src/uu/stty/src/stty.rs:266]: Ah, that's right, indeed they should be printed if they differ from default. I'll put a TODO, as those `min` and `time` special cases are a headache

File: src/uu/stty/src/stty.rs (hunk 1 above)
Comment (dezgeg) [reply src/uu/stty/src/stty.rs:262]: Sounds good, esp with `prefix` -> `meta_prefix` and `ctrl` -> `ctrl_prefix`

Comment (tertsdiepraam) [reply src/uu/stty/src/stty.rs:262]: Yes, that's even better!

File: src/uu/stty/src/stty.rs (hunk 2 above)
Comment (tertsdiepraam) [reply src/uu/stty/src/stty.rs:266]: Fair enough!

File: src/uu/stty/src/stty.rs [hunk 3]
```diff
@@ -245,6 +245,47 @@ fn print_terminal_size(termios: &Termios, opts: &Options) -> nix::Result<()> {
...
+        ("", cc)
+    };
+
+    let (ctrl_prefix, character) = match code {
+        0..=0x1f => Ok(("^", (b'@' + code) as char)),
+        0x20..=0x7e => Ok(("", code as char)),
+        0x7f => Ok(("^", '?')),
+        _ => Err(nix::errno::Errno::ERANGE),
+    }?;
+
```
Comment (sylvestre) [inline src/uu/stty/src/stty.rs:265]:
```suggestion
    // Determine the prefix and character based on the code
    let (ctrl_prefix, character) = match code {
        // Control characters in ASCII range
        0..=0x1f => Ok(("^", (b'@' + code) as char)),
        // Printable ASCII characters
        0x20..=0x7e => Ok(("", code as char)),
        // DEL character
        0x7f => Ok(("^", '?')),
        // out of range
        _ => Err(nix::errno::Errno::ERANGE),
    }?;

```

File: src/uu/stty/src/flags.rs [hunk 4]
```diff
@@ -313,3 +317,22 @@ pub const BAUD_RATES: &[(&str, BaudRate)] = &[
...
+    #[cfg(target_os = "linux")]
+    ("swtch", S::VSWTC),
+    ("start", S::VSTART),
+    ("stop", S::VSTOP),
+    ("susp", S::VSUSP),
+    ("rprnt", S::VREPRINT),
+    ("werase", S::VWERASE),
+    ("lnext", S::VLNEXT),
+    ("discard", S::VDISCARD),
+];
```
[the remark this answers is not in the document]
Comment (dezgeg) [reply src/uu/stty/src/flags.rs]: Good comments, squashed in.

File: src/uu/stty/src/flags.rs [hunk 5]
```diff
@@ -313,3 +317,41 @@ pub const BAUD_RATES: &[(&str, BaudRate)] = &[
     ))]
     ("4000000", BaudRate::B4000000),
 ];
+/// Control characters for the stty command.
+///
+/// This constant provides a mapping between the names of control characters
```
Comment (sylvestre) [inline src/uu/stty/src/flags.rs:322]: needs a rustfmt

Comment (dezgeg) [reply src/uu/stty/src/flags.rs:322]: Oops, forgot to `git add`... should be there now

**Comment by sylvestre** [comment]
is it too hard to write tests for these?

**Comment by dezgeg** [comment]
> is it too hard to write tests for these?

Main problem is in the existing tests there isn't anything suitable to start modifying; existing cases either test fails that happen in argument parsing stage or are not run due to `#[ignore = "Fails because cargo test does not run in a tty"]`

**Pull request merged**